Add GroupBy option to snapshots command

ArcticXWolf opened the pull request:
What is the purpose of this change? What does it change?
--------------------------------------------------------
This commit adds a --group-by option to the snapshots command, which
behaves similar to the --group-by option of forget. Valid option values
are "host, paths, tags". If this option is given, the output of
snapshots will be divided into multiple tables, according to the value
given (i.e. "host" will create a table of snapshots for each host, that
has a snapshot in the list). Also the JSON output will be grouped.

The default behavior (when --group-by is not given) has not changed.

More to this discussion can be found in issue #2037.

Was the change discussed in an issue or in the forum before?
------------------------------------------------------------
Yes, it was discussed in issue #2037.

Closes #2037

Checklist
---------

- [x] I have read the [Contribution Guidelines](https://github.com/restic/restic/blob/master/CONTRIBUTING.md#providing-patches)
- [x] I have added tests for all changes in this PR
- [x] I have added documentation for the changes (in the manual)
- [x] There's a new file in `changelog/unreleased/` that describes the changes for our users (template [here](https://github.com/restic/restic/blob/master/changelog/TEMPLATE))
- [x] I have run `gofmt` on the code in all commits
- [x] All commit messages are formatted in the same style as [the other commits in the repo](https://github.com/restic/restic/blob/master/CONTRIBUTING.md#git-commits)
- [x] I'm done, this Pull Request is ready for review

ArcticXWolf committed:
Add GroupBy option to snapshots command

This commit adds a --group-by option to the snapshots command, which
behaves similar to the --group-by option of forget. Valid option values
are "host, paths, tags". If this option is given, the output of
snapshots will be divided into multiple tables, according to the value
given (i.e. "host" will create a table of snapshots for each host, that
has a snapshot in the list). Also the JSON output will be grouped.

The default behavior (when --group-by is not given) has not changed.

More to this discussion can be found in issue #2037.

ArcticXWolf commented:
Hey, I want to work on the group-by option discussed in #2037. The main functionality is already done, whats missing are tests, documentation and changelogs.

As this is my first contribution to restic and also my first time coding in Go, I have some questions:

1. What's used to do the testing? (As I'm coming from Ruby, I know that there are many testing frameworks/DSLs)
2. Is the style of my changes okay? I'm not sure, if I violated any naming conventions or else.

I hope my contribution helps!

fd0 commented on cmd/restic/cmd_snapshots.go:
```diff
@@ -61,28 +70,87 @@ func runSnapshots(opts SnapshotOptions, gopts GlobalOptions, args []string) erro
 		}
 	}
 
+	// group by hostname and dirs
```
This is duplicate code, I would suggest extracting it into a function which takes a list of snapshots (`[]restic.Snapshot`, called a "slice" in Go) and returns a `map[string][]Snapshot` and an `error`. Let me know if you need help with this.

ArcticXWolf replied:
Okay, I extracted it into an own function. Is the style of the new function okay? And shall I remove the duplicate code (in the forget command) in this pull request or shall I open a new one after this got merged?

ArcticXWolf replied:
I put the shared code into the internal folder and made the forget command use it too.

fd0 commented on cmd/restic/cmd_snapshots.go:
```diff
@@ -223,6 +291,40 @@ func PrintSnapshots(stdout io.Writer, list restic.Snapshots, reasons []restic.Ke
+// following snapshots belong to.
+// Prints nothing, if we did not group at all.
+func PrintSnapshotGroupHeader(stdout io.Writer, groupKeyJSON string, GroupByTag bool, GroupByHost bool, GroupByPath bool) error {
+	if GroupByTag || GroupByHost || GroupByPath {
```
It's common in Go to return early, so the actual action in the function is not indented so far. How about this instead?
```suggestion
	if !GroupByTag && !GroupByHost && !GroupByPath {
		return nil
	}
```

fd0 commented on cmd/restic/cmd_snapshots.go:
```diff
@@ -231,20 +333,60 @@ type Snapshot struct {
 
+// SnapshotGroup helps to print SnaphotGroups as JSON with their GroupReasons included.
+type SnapshotGroup struct {
+	GroupKey  groupKey
```
Add so-called struct tags here to set the names in the JSON output
```suggestion
	GroupKey  groupKey `json:"group_key"`
```

fd0 commented on cmd/restic/cmd_snapshots.go:
```diff
@@ -231,20 +333,60 @@ type Snapshot struct {
+// SnapshotGroup helps to print SnaphotGroups as JSON with their GroupReasons included.
+type SnapshotGroup struct {
+	GroupKey  groupKey
+	Snapshots []Snapshot
```
```suggestion
	Snapshots []Snapshot `json:"snapshots"`
```

fd0 commented on cmd/restic/cmd_snapshots.go:
```diff
@@ -45,6 +47,13 @@ func init() {
+}
+
+type groupKey struct {
+	Hostname string
```
Add so-called struct tags here to set the name for the JSON output (lowercase, snake_case)
```suggestion
	Hostname string `json:"hostname"`
```

fd0 reviewed (changes requested):
Hey, I've had a look at the code, thanks for your contribution (especially as a first-time contributor new to Go)!

I've got a few things that you could improve in the code, but it looks great already.

 * The Tests on Travis fail because the file `internal/archiver/testing_test.go` is not formatted properly with `gofmt` (the Go formatting standard is an implementation, rather than a document), which has changed a bit recently in the Go 1.11 release. In order to pass the tests, you need to format the code with `gofmt` on Go 1.11.
 * There are modifications to files in `vendor`/, these should not be there. Did you maybe run `gofmt` on all files, including `vendor/`? Please remove the modifications from the PR.

For testing, we're using the builtin testing framework, you can easily run all tests with `go test ./...` in the repo, or chdir to some directory and just run `go test`.

Unfortunately, we don't have tests for user-interface interactions like this in place (at least not yet).

Let me know if you need any help!

ArcticXWolf committed:
Fix json tags for grouped snapshot output

This commit will add json tags to the structs for json output, so all
json variables of the snapshot command output are lowercase and
snake-case.

Furthermore it adds some internal code changes based on the feedback in
the pull request #2087.

ArcticXWolf committed:
Move snapshot grouping code into own function to deduplicate code

This commit moves the code which is used to group snapshots in the
snapshots command into an own function to deduplicate code shared by the
snapshots command and forget command.

ArcticXWolf committed:
Add changelog of group-by option for snapshots command

ArcticXWolf commented:
Okay, this is nearly complete, I just have some last questions:

> For testing, we're using the builtin testing framework, you can easily run all tests with go test ./... in the repo, or chdir to some directory and just run go test.
> 
> Unfortunately, we don't have tests for user-interface interactions like this in place (at least not yet).

The tests succeed with my changes, but shall I add new ones? I have no idea how I should design such UI tests. Maybe just save the expected output into a fixture?

Furthermore I added a changelog file, whats missing now is the modification of the documentation. Where shall I mention it? Maybe at [the working with repos section](https://github.com/restic/restic/blame/master/doc/045_working_with_repos.rst#L56)?

ArcticXWolf committed:
Refactor duplicate code for grouping snapshots

This commit is a followup to the addition of the --group-by flag for the
snapshots command. Adding the grouping code there introduced duplicated
code (the forget command also does grouping). This commit refactors
boths sides to only use shared code.

ArcticXWolf committed:
Add documentation for --group-by flag for snapshots command

ArcticXWolf commented:
Okay, I refactored the forget command to use the shared code and added the documentation. Tests are green for me, just waiting for the CI to confirm this. As I do not know how to do the interface tests (and we did not have them before), I believe this is ready for review.

fd0 reviewed (approved):
LGTM, thanks! Sorry for the long delay before reviewing this!

fd0 commented:
I've rebased and fixed the conflict on `cmd_forget.go`, let's see what the tests say.
